edx_django_utils/monitoring/internal/backends.py

```diff
@@ -0,0 +1,177 @@
+        """
+
+
+class NewRelicBackend(TelemetryBackend):
+    """
+    Send telemetry to New Relic.
+
+    https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/guide-using-python-agent-api/
+    """
+    def __init__(self):
+        if newrelic is None:
+            raise Exception("Could not load New Relic monitoring backend; package not present.")
+
+    def set_attribute(self, key, value):
+        # Sets attribute on the transaction, rather than the current
+        # span, matching historical behavior. There is also an
+        # `add_custom_span_attribute` that would better match
+        # OpenTelemetry's behavior, which we could try exposing
+        # through a new, more specific TelemetryBackend method.
+        #
+        # TODO: Update to newer name `add_custom_attribute`
+        # https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/addcustomparameter-python-agent-api/
+        newrelic.agent.add_custom_parameter(key, value)
+
+    def record_exception(self):
+        # TODO: Replace with newrelic.agent.notice_error()
+        # https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/recordexception-python-agent-api/
+        newrelic.agent.record_exception()
+
+
+class OpenTelemetryBackend(TelemetryBackend):
+    """
+    Send telemetry via OpenTelemetry.
+
+    Requirements to use:
+
+    - Install `opentelemetry-api` Python package
+    - Configure and initialize OpenTelemetry
+
+    API reference: https://opentelemetry-python.readthedocs.io/en/latest/
+    """
+    # pylint: disable=import-outside-toplevel
+    def __init__(self):
+        # If import fails, the backend won't be used.
+        from opentelemetry import trace
+        self.otel_trace = trace
+
+    def set_attribute(self, key, value):
+        # Sets the value on the current span, not necessarily the root
+        # span in the process.
+        self.otel_trace.get_current_span().set_attribute(key, value)
+
+    def record_exception(self):
+        self.otel_trace.get_current_span().record_exception(sys.exc_info()[1])
+
+
+class DatadogBackend(TelemetryBackend):
+    """
+    Send telemetry to Datadog via ddtrace.
+
+    Requirements to use:
+
+    - Install `ddtrace` Python package
+    - Initialize ddtrace, either via ddtrace-run or ddtrace.auto
+
+    API reference: https://ddtrace.readthedocs.io/en/stable/api.html
+    """
+    # pylint: disable=import-outside-toplevel
+    def __init__(self):
+        # If import fails, the backend won't be used.
+        from ddtrace import tracer
+        self.dd_tracer = tracer
+
+    def set_attribute(self, key, value):
+        if root_span := self.dd_tracer.current_root_span():
+            root_span.set_tag(key, value)
+
+    def record_exception(self):
+        if span := self.dd_tracer.current_span():
+            span.set_traceback()
+
+
+# This has to be cached rather than computed on load because otherwise
+# we get a (silent) circular import.
```

Comment:
[curious passerby] Could you explain this comment? I think that the circularity here is that the execution will end up importing the same file(through `import_string`) but I don't see how that's a problem.

Reply:
The usual alternative to using a cache decorator would be something like `_BACKENDS = configured_backends()` at the top level of the module. That would fail because it would try to `import_string('edx_django_utils.monitoring.NewRelicBackend')` while the module itself was still being loaded, resulting in an ImportError. The error would be caught and the backend would be dropped from the list, which is why I referred to it as silent.

Maybe I can rephrase the comment a bit. perhaps something like this?

```python
# We're using an lru_cache instead of assigning the result to a
# variable on module load.  With the default settings (pointing to a
# TelemetryBackend in this very module), this function can't be
# successfully called until the module finishes loading, otherwise we
# get a circular import error that will cause the backend to be
# dropped from the list.
```

Reply:
I like this explanation more, as it is more clear and both zawan-ila and i were a little confused.